core/src/main/java/org/elasticsearch/index/engine/InternalEngine.java

```diff
@@ -650,14 +650,24 @@ private IndexResult innerIndex(Index index) throws IOException {
                 }
                 updatedVersion = index.versionType().updateVersion(currentVersion, expectedVersion);
                 index.parsedDoc().version().setLongValue(updatedVersion);
+
+                // Update the document's sequence number and primary term, the
+                // sequence number here is derived here from either the sequence
+                // number service if this is on the primary, or the existing
+                // document's sequence number if this is on the replica. The
+                // primary term here has already been set, see
+                // IndexShard.prepareIndex where the Engine.Index operation is
+                // created
+                index.parsedDoc().updateSeqID(seqNo, index.primaryTerm());
+
                 if (currentVersion == Versions.NOT_FOUND && forceUpdateDocument == false) {
                     // document does not exists, we can optimize for create, but double check if assertions are running
                     assert assertDocDoesNotExist(index, canOptimizeAddDocument == false);
                     index(index.docs(), indexWriter);
                 } else {
                     update(index.uid(), index.docs(), indexWriter);
                 }
-                indexResult = new IndexResult(updatedVersion, seqNo, deleted);
+                indexResult = new IndexResult(updatedVersion, seqNo, index.primaryTerm(), deleted);
```

Comment:
I see. The seqNo and the term do not necessarily always go together. the seqNo is the location of the operation and the term is the authority to put in there. I like the fact that the result object only contains the things that the internal engine creates / changes. Seq# are owned by the engine (on a primary). Terms are owned by the shard. I would prefer to remove the term. At least in the example you gave (`Translog.Index#Index(Index,  IndexResult`) it's readily available from the index operation.